src/librustc_codegen_ironox/x86_asm_printer.rs

```diff
@@ -0,0 +1,79 @@
+use super::ModuleIronOx;
+
+use context::CodegenCx;
+use value::{Value, Instruction};
+
+use rustc::mir::mono::Stats;
+use rustc::util::nodemap::FxHashMap;
+use rustc_codegen_ssa::traits::MiscMethods;
+use std::cell::{Cell, RefCell};
+
+/// A convenience macro for appending instructions to a string.
+macro_rules! asm {
+    (
+        $m:ident,
+        $(
+            $fmt:expr; [ $($args:expr),* ]
+        ),*
+    ) => {
+        let mut asm_lines = vec![$(format!($fmt, $($args),*)),*];
+        asm_lines = asm_lines.iter().map(|x| format!("\t{}\n", x)).collect();
+        let asm_lines = asm_lines.join("");
+        $m.push_str(&asm_lines);
+    };
+    ($m:ident, $new_asm:expr) => {
+        $m.push_str($new_asm);
+    }
+}
+
+/// A convenience macro for appending a label to a string.
```

Comment:
So these macros...

I'm wondering if it's necessary. Your label macro could be:
```
fn label(m: &mut Vec<Str>, lbl: &str) {
    m.push(&format!("{}:\n", lbl));
}
```

Reply:
The `asm` macro makes it a bit easier to append lines to the `asm` string (which contains the compilation result). I don't think I can make `asm` a function and achieve the same result. For example, I might the `asm` macro can be used to append instructions to the `asm` string like this:

```rust
                    asm!(asm, "mov {}, %rax"; [dest],
                              "movq {}, %rbx"; [source],
                              "movq %rbx, (%rax)"; []);
                    label!(asm, "some_label");
```

I use both `asm` and `label` to manipulate the `asm` string, so I decided `label` might as well be a macro too. I guess I was aiming for consistency. Do you think `label` should be a function instead?

(Laurie mentioned I should probably remove the macros from this PR, since they're unused anyway, so I will address your comment in my next PR).

Reply:
If there's a good reason for the macros (in the next PR) then I don't object to keeping them.

Reply:
I agree that the macros make things easier. One thing to maybe think about for the next PR: if we always append to `asm`, we could remove that as an argument to the macro?

Reply:
That's a fair point.